include/drivers/rtc/ds3231.h

```diff
@@ -0,0 +1,398 @@
+/*
```

Comment:
i wonder about the location of this file. implementation is in `drivers/counter`. What about putting it in `include/drivers/counter`?

Reply:
That's somewhat specific to the example, but not entirely, so I'll outline the general issue.

Fundamentally the DS3231 is a real-time-clock device, not a counter.  It's a counter only because that's the closest existing generic driver class.  The source is in `counter` because it implements that API, but anybody who wants to use it will be operating near exclusively with the device-specific functions.

IMO using `#include <drivers/counter/ds3231.h>` will mislead people into thinking it really is a counter.  But DS3231 can support only extremely limited use of the counter API, and some things that are expected (like being able to read from within an ISR) are totally unsupported.  A strong case can be made that it shouldn't be a counter at all.

But that opens up the question of where to put devices that have no relation to any generic driver class.  Or, to whether in this case we should bring back an RTC driver API that provides the functionality required for using an RTC separately from counter.

Discussion about that probably belongs in the PR for the DS3231 module though.  In the other driver for this sort of thing--extensions to the sensor API--the devices fully support the base driver class, and simply require a few extra functions to expose their full functionality.  Those would go into `include/drivers/sensors/abc321z.h`.

Reply:
i agree, that this case is specific. In case of driver which fits to specific api and extends it, there will be no such problem.

Reply:
Implementing the generic counter interface will make people think it's a counter. If you're worried about people thinking it's a counter, then why implement the generic counter interface at all?

Similarly, if you think applications will use the device-specific functions exclusively, then isn't implementing the generic counter interface a waste of code space? I think we either need to revisit the RTC driver API or you should have a driver-specific API.

That said, I think this is nice way to extend a generic driver API if that functionality is needed. I'm just not sure that DS3231 specifically is a good use case for it.